tutorials/haskell/generics/tutorial.md

```diff
@@ -0,0 +1,584 @@
+kind `*`, such as `Show`, but if we want to use this system to derive a
+`Functor` instance, we need to allow it to work with kind `* -> *`.
+
+We want to be able to derive `Functor` (and similar) instances generically,
+so we have to fix things up. The solution is to add one more type parameter
+`p` to all our combinators:
+
+```haskell
+data V1 p
+data U1 p = U1
+data (f :+: g) p = L1 (f p) | R1 (g p)
+data (f :*: g) p = (f p) :*: (g p)
+```
+
+This way `fmap` would be:
+
+```haskell
+fmap :: (a -> b) -> Rep f a -> Rep f b
+--                  p  =  a    p  =  b
+```
+
+But what happens to the type classes that work with `*` kinds? Here are our
+choices:
+
+* Have a separate set of combinator types for each case (`*` and `* -> *`
+  kinds).
+
+* Use the most general form (with `p`), but for `*` kinds just treat the
+  extra `p` parameter as a dummy type index that has no meaning.
+
+The authors of the generic extension went with the second option, and I
+can't blame them. We'll see that there are a lot of wrappers already and we
+really should try to keep their number from exploding any further.
+
+Let's try to map from a data type to its representation and see if we're
+still missing something:
+
+```haskell
+data Maybe a = Nothing | Just a
+
+-- Interestingly, we could build a representation that works on ‘Maybe a’,
+-- that is, a thing of kind *, if we wanted to derive something like ‘Show’.
+-- At the same time if we wanted to derive ‘Functor’, we would work with
+-- ‘Maybe’ of kind * -> *. This means that there are actually two different
+-- possible representations depending on our aim. This is addressed with two
+-- different generics type classes, as we will see later.
+
+-- For kind *, things like ‘Show’:
+
+-- type: (U1 :+: ?) p
+```
+
+Oh, how to represent `Just a`? We need to way to allow it have an argument!
+Let's add the following:
+
+```haskell
+data Rec0 c p = Rec0 { unRec0 :: c }
+```
+
+`Rec` part in the type's name hints that this thing may be possibly
+recursive.
+
+But I'm lying to you, due to some historical stuff, it's defined a bit
+differently:
+
+```haskell
+type Rec0 = K1 R
+
+newtype K1 i c p = K1 { unK1 :: c } -- c is the value, ‘a’ in ‘Maybe a’
+--         ^   ^
+--         |   |
+--         |   +-------- dummy p
+--         |
+-- type-level tag, R or P
+```
+
+You see this `R` type-level tag? There used to be another one, `P` and this
+`type Par0 = K1 P` type synonym, but forget it, it's deprecated. Bottom
+line: `Rec0` is used for data constuctor arguments (fields), that **are
+not** `p` parameter.
+
+With `Rec0`, we can finally build representation of `Maybe a`:
+
+```haskell
+-- This is type of our representation: (U1 :+: Rec0 a) p
+
+-- Examples of values for ‘Maybe Int’:
+
+-- Nothing => L1 U1
+-- Just 5  => R1 (K1 5) -- remember where K1 comes from?
+--            ^
+--            |
+--            +--- L1 and R1 are from our representation of sum types
+```
+
```

Comment:
Could we add in a sentence or two explaining the motivation for coming up with a second representation of Maybe?  Also, which representation is the one actually used by ghc?  Or are both used depending on context?